### ISSUE TITLE
ci: more visibility of errors output by tools/check.sh

### ISSUE THREAD

**Issue by johelegp** [cplusplus/draft#4529]
@burblebee commented at https://github.com/cplusplus/draft/pull/4520#issuecomment-786393620 about wanting something like that. Does the following meet your expectations? https://github.com/johelegp/draft/runs/1985020607?check_suite_focus=true#step:4:2160. Maybe what you actually want is a bot that comments the errors on the PR. Or have the errors output in a separate step, so that you don't have to scroll all the LaTeX build output.

**Comment by jwakely** [comment]
I was about to open an issue about exactly this.

Why doesn't the script just print "Checking foo..." before each check, so if there's output from a check you know where it comes from? Seems easy. Or if you want to keep the output empty when nothing fails, make those headers only printed with a --verbose switch (so we can at least run it locally to find the problem). Or filter the output of failed checks through `sed 's/^/SOME IDENTIFIYING TOKEN/'`

There are lots of ways to make this less abstruse.

**Comment by jwakely** [comment]
Adding "Error: " is better than nothing, but for many of the checks in that script there is no context at all. The "non-ASCII character" output is clear enough for me, but there is no context for checks for things like `\returns` not being followed by a newline.

**Comment by jwakely** [comment]
For example, the failure at https://github.com/cplusplus/draft/runs/1978623293?check_suite_focus=true just ends like this:

```
texlive/texmf-dist/fonts/type1/public/rsfs/rsfs10.pfb>
Output written on std.pdf (1849 pages, 7045850 bytes).
Transcript written on std.log.
Latexmk: Log file says output to 'std.pdf'
Latexmk: All targets (std.pdf) are up-to-date
utilities.tex:584:\returns \tcode{static_cast<underlying_type_t<T>>(value)}.
Error: Process completed with exit code 1.
```

I'm pretty familiar with the draft sources and the conventions, but that had me stumped until I read `tools/check.sh` in detail.

This change:

**Comment by jwakely** [comment]
(Sorry for closing it, I tabbed my way to the close button and hit enter :facepalm:)

**Comment by jwakely** [comment]
Would it make sense to run the check script **first** (because it's really fast) so if the document builds but fails a lint check you don't have to scroll past thousands of lines of output from latex to find the error?

If a lint check is going to fail the CI run anyway, why bother building the document? It's probably safe to assume the person making the change has built it, even if they didn't run the linter checks. So building it first adds no value.

**Comment by jwakely** [comment]
And maybe the CI build could also say "Running tools/check.sh" before running it, so it's more obvious where that error comes from. More context is good.

**Comment by jwakely** [comment]
I'm not sure why the install step is failing in the CI checks for my pull request. Does my fork need something enabled?

**Comment by johelegp** [comment]
> **Edit:** looks like we can't run it earlier, because the script depends on some of the build artefacts created during the `make` step. Maybe we want two scripts then, or a "quick check" mode for the script, so we can run some of it first to fail fast, then do the build and re-check.

Seems worth it. It could even run before the "install" step, which takes ~40 seconds, compared to the ~30 seconds from the check script (which I got by comparing a failed run vs a successful run. We should get the exact time once it's in its own step).

> I'm not sure why the install step is failing in the CI checks for my pull request. Does my fork need something enabled?

Seems like all your runs were done on Ubuntu 20.04, which doesn't have the texlive-generic-recommended package.

**Comment by burblebee** [comment]
Issues I have with the checker script:
*it doesn't work on iMacs due to Apple's broken sed (I had to modify the script locally to remove the checks that don't work on the mac, but it means I can't run those extra checks before committing)
*it doesn't check for embedded tabs (the travis checks did)
*not all errors are flagged red (it's the only sensible way to find errors in the reams of build output)

**Comment by jwakely** [comment]
We can probably make the use of GNU sed features optional, or replace them with POSIX alternatives.

The CI actions use `make quiet` but it's not quiet at all. I think that's because it has to repeat the build to create the cross references, and the repeated build doesn't use `-silent` .

**Comment by johelegp** [comment]
Yes. It is intentionally only quiet only for the first pass: https://github.com/cplusplus/draft/pull/3775/commits/e9563c295bbdacfafe3e313513ba2ecbbcddbe1e.

**Comment by jensmaurer** [comment]
superseded by #4533

**Comment by johelegp** [comment]
I think you meant to supersede #4530 and resolve this with the word of power Resolves #4529.

**Comment by burblebee** [comment]
> We can probably make the use of GNU sed features optional, or replace them with POSIX alternatives.

Making them optional doesn't help me - I want to be able to run all checks to catch errors before committing.
